thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomalydetection/model/transform/AnomalyRemovalFunction.java

```diff
@@ -0,0 +1,180 @@
+   */
+  @Override
+  public TimeSeries transform(TimeSeries timeSeries, AnomalyDetectionContext anomalyDetectionContext) {
+    Interval timeSeriesInterval = timeSeries.getTimeSeriesInterval();
+    long startTime = timeSeriesInterval.getStartMillis();
+    long endTime = timeSeriesInterval.getEndMillis();
+
+    long bucketSizeInMillis = anomalyDetectionContext.getBucketSizeInMS();
+
+    // get monitoring window
+    String metricName = getProperties().getProperty(TARGET_METRIC_NAME, DEFAULT_TARGET_METRIC_NAME);
+    String metricTimezone = getProperties().getProperty(METRIC_TIMEZONE, DEFAULT_METRIC_TIMEZONE);
+    Interval currentWindow = anomalyDetectionContext.getCurrent(metricName).getTimeSeriesInterval();
+
+    // get historical anomalies
+    List<MergedAnomalyResultDTO> anomalyHistory = anomalyDetectionContext.getHistoricalAnomalies();
+
+    double weightThreshold =
+        Double.valueOf(getProperties().getProperty(ANOMALY_REMOVAL_WEIGHT_THRESHOLD, DEFAULT_ANOMALY_REMOVAL_WEIGHT_THRESHOLD));
+    int toleranceSize =
+        Integer.valueOf(getProperties().getProperty(ANOMALY_REMOVAL_TOLERANCE_SIZE, DEFAULT_ANOMALY_REMOVAL_TOLERANCE_SIZE));
+
+    TimeSeries transformedTimeSeries = new TimeSeries();
+    Interval transformedInterval = new Interval(startTime, endTime);
+    transformedTimeSeries.setTimeSeriesInterval(transformedInterval);
+
+    Map<Long, Boolean> anomalousTimestamps =
+        getAnomalousTimeStampAndLabel(anomalyHistory, bucketSizeInMillis, weightThreshold);
+
+    long windowStartOffset =
+        getOffsetTimestamp(currentWindow.getStartMillis(), toleranceSize, bucketSizeInMillis, metricTimezone);
+
+    // generate the transformed time series
+    for (long ts:timeSeries.timestampSet()) {
+      Boolean userLabel = anomalousTimestamps.get(ts);
+      userLabel = (userLabel == null) ? false : userLabel;
+      if (ts < windowStartOffset || (userLabel && ts < currentWindow.getStartMillis())) {
+        transformedTimeSeries.set(ts, Double.NaN);
+      } else {
+        transformedTimeSeries.set(ts, timeSeries.get(ts));
+      }
+    }
+
+    return transformedTimeSeries;
+  }
+
+  /**
+   * Extract the timestamps and their labels of historical anomalies if
+   *  1) user labeled the anomaly as a true anomaly
+   *  2) without label and the anomaly severity surpass the default weight (severity indicator) threshold
+   *     this only works when a s
+   * @param historicalAnomalies anomaly results in the past
+   * @param bucketMillis the bucket in terms of milliseconds
+   * @param weightThreshold the threshold to use on weight if user label is not given
+   * @return the Map of timestamps for anomalies and their labels
+   */
+  public static Map<Long, Boolean> getAnomalousTimeStampAndLabel(
+      List<MergedAnomalyResultDTO> historicalAnomalies, long bucketMillis, double weightThreshold) {
+
+    Map<Long, Boolean> anomalousTimestamps = new HashMap<>();
+    if (!historicalAnomalies.isEmpty()) {
```

Comment:
This line is redundant. You may need to check if historicalAnomalies is null. As you have for each in the next line, you don't need to check if it is empty.